### ISSUE TITLE
md6 implementation in pure javascript

### ISSUE THREAD

**Issue by Kikobeats** [sindresorhus/module-requests#37]
why not!

https://groups.csail.mit.edu/cis/md6/

[C implementation reference](https://groups.csail.mit.edu/cis/md6/downloads.html)

**Comment by sindresorhus** [comment]
Doesn't look like it's used very widely. https://en.wikipedia.org/wiki/MD6

**Comment by forivall** [comment]
Well, adoption of hash functions is typically quite slow, so, given that it was first published in 2008, I wouldn't expect wide usage.

Would you expect a true pure javascript implementation, or a pure node implementation, or an implementation limited to use of ArrayBuffer or Node Buffer?

**Comment by yumitsu** [comment]
@Kikobeats see https://github.com/Snack-X/md6/blob/master/md6.js

**Comment by Kikobeats** [comment]
@sindresorhus not really, but md5 algorithm is popular and presents hash collisions since 1996.

@forivall actually I don't know. I think that the goal if this library should be create a hash in the less time possible. Maybe could be sense use something with node-gyp or from a derived library to compile the original C code into a JS library? what do you think?

@yumitsu Looks good at starting point! 🙃

**Comment by yumitsu** [comment]
@Kikobeats I think we can extract this source to standalone package and modify it (thanks to license).

**Comment by sindresorhus** [comment]
@Kikobeats Yes, that's why we have SHA256 and SHA512.

Crypto is sensitive, though, so I wouldn't trust some random code out there with a WTFPL "license".

**Comment by Kikobeats** [comment]
@sindresorhus totally true. This is the reason why I opened this module request. But now I'm thinking in convert directly the official C code into a JS library (time to investigate how to), so feel free to close the issue :-)

**Comment by sindresorhus** [comment]
You can use [emscripten](https://github.com/kripken/emscripten) to compile it to highly optimized JS.

@rodneyrehm has used emscripten to compile [Sass to JS](https://github.com/medialize/sass.js), so he might be of help.

**Comment by Kikobeats** [comment]
Perfect! 🙌

**Comment by rodneyrehm** [comment]
Isn't the [Web Crypto API](http://www.w3.org/TR/WebCryptoAPI/) the place for these things?

Making a C/C++ library available to node is best done via node-gyp (The node-sass people, e.g. @xzyfer have experience with this). It is likely *much* faster than C/C++ turned JavaScript. Making a C/C++ library available to the browser can be done via emscripten. Considering MD6 doesn't have a lot of fancy things going on (like calling user space JS, async execution, …) this should be comparably simple.

---

Having had a look at md6.h, you'll need to create "simplification wrappers" for `md6_hash()` and `md6_full_hash()`. Because emscripten can only handle primitive types, you'll have to accept `int` and cast that to `uint64_t` in C space. To avoid stack-and-pointer-fuckery-from-JS, you'll want to use [`EM_ASM_INT()`](https://kripken.github.io/emscripten-site/docs/api_reference/emscripten.h.html#c.EM_ASM_INT) in your C wrapper to export the generated `unsigned char *hashval` back to JS: 

```C
// passes pointer to JS land, where we use
// Module.Pointer.stringify() to convert pointer to text
const char* dummy = strdup("hello world");
EM_ASM_INT({
  var text = $0 && Module.Pointer_stringify($0) || null;
  console.log("got string from C", text);
}, dummy);
```

To make this work, your `md6_emscripten_wrapper.c` will have to `#include <emscripten.h>`.

You can also expose `md6_init()` et al for chunked hashing. emscripten does not play too well with structs, so the `md6_state` has to be created and maintained in C space. You can pass a pointer to that state back and forth without a problem.

---

The JS generated by emscripten can run in the browser's main thread or in a WebWorker. Of course the JS can also run in node, rhino, nashorn, etc. Do not run a minifier (like UglifyJS) yourself, it will destroy the ASM comments, resulting in warnings - and slower execution - in Firefox. Emscripten's runtime itself is somewhere around 600KB of (unminified) JS, so be prepared for your MD6 to not come cheap, bandwidth wise.

---

Let me know if you need assistance. Good Luck :)

**Comment by Qix-** [comment]
If this is node-specific, and you want to use C++, I recommend against Emscripten as painful as C++ bindings are. Use Nan, and wrapping a few C-functions should be relatively easy.

**Comment by Richienb** [comment]
@Kikobeats Does https://github.com/Neo-Desktop/md6 match your use case?
If so, @sindresorhus can close this issue.

**Comment by sindresorhus** [comment]
@Richienb I wouldn't trust a random fork that hasn't been updated in years.

**Comment by Richienb** [comment]
@sindresorhus In that case, I'll fork the original source and roll my own npm package.

**Comment by sindresorhus** [comment]
I honestly still wouldn't trust that the MD6 implementation is correct. Neither the original repo or the fork look very professional. And the test suite is not very comprehensive.

**Comment by Richienb** [comment]
@sindresorhus Emscripten it is!

**Comment by Richienb** [comment]
Emscripten was making pretty large wasm files so I ended up adapting https://github.com/Snack-X/md6 which seems to be made by someone who has been making javascript implementations of cryptographic functions for quite some time now.

Taking the source code, I resolved the issue from the original code - most of which were problems with javascript function equalivents and made it more javascripty.

https://github.com/Richienb/md6-hash
